src/projection.jl

```diff
@@ -52,93 +128,197 @@ function ProjectTo(x::T) where {T<:Union{<:Tuple,NamedTuple}}
-(::ProjectTo{T})(dx::AbstractZero) where {T} = zero(T)
+(::ProjectTo{T})(dx::AbstractZero) where {T} = dx # zero(T)
+
+# Thunks
+(project::ProjectTo)(dx::Thunk) = Thunk(project ∘ dx.f)
+(project::ProjectTo)(dx::InplaceableThunk) = Thunk(project ∘ dx.val.f) # can't update in-place part
+(project::ProjectTo)(dx::AbstractThunk) = project(unthunk(dx))
 
-# Number
-ProjectTo(::T) where {T<:Number} = ProjectTo{T}()
+# Non-differentiable -- not so sure this is necessary. Keeping value like this is a bit awkward.
+for T in (:Bool, :Symbol, :Char, :String, :Val, :Type)
+    @eval ProjectTo(dx::$T) = ProjectTo{AbstractZero}(; value=dx)
+end
+(::ProjectTo{AbstractZero})(dx::AbstractZero) = dx
+(::ProjectTo{AbstractZero})(dx) = ZeroTangent()
+
+# Number -- presently preserves Float32, but not Int, is this ideal?
+ProjectTo(::T) where {T<:Number} = ProjectTo{float(T)}()
 (::ProjectTo{T})(dx::Number) where {T<:Number} = convert(T, dx)
 (::ProjectTo{T})(dx::Number) where {T<:Real} = convert(T, real(dx))
 
+ProjectTo(::Type{T}) where T<:Number = ProjectTo(zero(T)) # maybe
+
 # Arrays 
-ProjectTo(xs::T) where {T<:Array} = ProjectTo{T}(; elements=map(ProjectTo, xs))
-function (project::ProjectTo{T})(dx::Array) where {T<:Array}
-    _call(f, x) = f(x)
-    return T(map(_call, project.elements, dx))
-end
-function (project::ProjectTo{T})(dx::AbstractZero) where {T<:Array}
-    return T(map(proj -> proj(dx), project.elements))
-end
-(project::ProjectTo{<:Array})(dx::AbstractArray) = project(collect(dx))
+# ProjectTo(xs::T) where {T<:Array} = ProjectTo{T}(; elements=map(ProjectTo, xs))
+# function (project::ProjectTo{T})(dx::Array) where {T<:Array}
+#     _call(f, x) = f(x)
+#     return T(map(_call, project.elements, dx))
+# end
+# function (project::ProjectTo{T})(dx::AbstractZero) where {T<:Array}
+#     return T(map(proj -> proj(dx), project.elements))
+# end
+# (project::ProjectTo{<:Array})(dx::AbstractArray) = project(collect(dx))
 
 # Arrays{<:Number}: optimized case so we don't need a projector per element
-function ProjectTo(x::T) where {E<:Number,T<:Array{E}}
-    return ProjectTo{T}(; element=ProjectTo(zero(E)), size=size(x))
+function ProjectTo(x::AbstractArray{T,N}) where {T<:Number,N}
+    T == Bool && return ProjectTo(false)
+    sub = ProjectTo(zero(T))
+    return ProjectTo{AbstractArray{eltype(sub),N}}(; element=sub, axes=axes(x))
 end
-(project::ProjectTo{<:Array{T}})(dx::Array) where {T<:Number} = project.element.(dx)
-function (project::ProjectTo{<:Array{T}})(dx::AbstractZero) where {T<:Number}
-    return zeros(T, project.size)
+function (project::ProjectTo{AbstractArray{T,N}})(dx::AbstractArray{S,M}) where {T,S,N,M}
+    dy = S <: T ? dx : broadcast(project.element, dx)
+    if axes(dy) == project.axes
+        return dy
+    else
+        for d in 1:max(N,M)
+            size(dy, d) == length(get(project.axes, d, 1)) || throw(DimensionMismatch("wrong shape!"))
+        end
+        return reshape(dy, project.axes)
```

Comment:
The size check 2 lines up is meant to restrict this to cases which are almost the same, like `ones(4,1)` vs `ones(4)`.

Reply:
Doesn't ending up with `ones(4,1)` when your input was a `Vector` mean the rule is just wrong and should be fixed?

Reply:
Maybe. But such bugs are super-common. 

If we don't want to fix them, then it might not be necessary to store the axes at all, which would simplify things.

Reply:
Here's a fairly typical example: a non-LinearAlgebra function where Julia treats an adjoint vector as being a one-row matrix. Making every rule worry about this seems like precisely the complexity that this global projection is supposed to help us avoid. 
```
julia> xs = [[i, i^2, i^3] for i in 1:4];

julia> reduce(vcat, xs')
4×3 Matrix{Int64}:
 1   1   1
 2   4   8
 3   9  27
 4  16  64

julia> gradient(ys -> sum(abs, reduce(vcat, ys)), xs')[1][1]
1×3 Matrix{Int64}:
 1  1  1
```

Reply:
> Making every rule worry about this seems like precisely the complexity that this global projection is supposed to help us avoid.

This isn't about avoiding complexity at all, it is about avoiding wrong answers.

Reply:
I know, what I'm saying is, there would be a lot of complexity involved in making the `vcat` rule figure this out. You would end up writing literally what this projector does, as a last step, in that and in many other rules. Less complexity to extract it out into one place, i.e., here.

Reply:
Hmm, ok